<a href="https://colab.research.google.com/github/sbb2002/Portfolio/blob/main/.study/Tensorflow/ML_ch4_multivar_with_latex_guide.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()

Instructions for updating:
non-resource variables are not supported in the long term


### * LaTex 사용법
> https://colab.research.google.com/notebooks/markdown_guide.ipynb

# **Multi-variables linear regression**

H(x1, x2, x3, ..., xn) = w1 * x1 + w2 * x2 + w3 * x3 + ...  + b

너무 복잡하다. 이것을 matrix로 만든다.

> $H_{(X)}$ = **XW**

>> = $\begin{pmatrix}
x_{11} & x_{12} & \cdots & x_{1n} \\
x_{21} & x_{22} & \cdots & x_{2n} \\
\vdots & \vdots & \ddots & \vdots \\
x_{n1} & x_{n2} & \cdots & x_{nn}
\end{pmatrix}$
$\begin{pmatrix}
w_{1} \\
w_{2} \\
\vdots \\
w_{n}
\end{pmatrix}$ 

>> = [n(instance), n(variables] X [n, 1] = [n(instance), 1(Y)]


이제 위 수식에 따라 **X** 를 입력하자.
 
| $x_{1}$ | $x_{2}$ | $x_{3}$ | Y |
| - | - | - | - |
| 73 | 80 | 75 | 152 |
| 93 | 88 | 93 | 185 |
| 89 | 91 | 90 | 180 |
| 96 | 98 | 100 | 196 |
| 73 | 66 | 70 | 142 |

In [ ]:
# Give the data above.
x1_data = [73,93,89,96,73]
x2_data = [80,88,91,98,66]
x3_data = [75,93,90,100,70]
y_data = [152,185,180,196,142]

# Placeholders for a tensor that will be always fed.
x1 = tf.placeholder(tf.float32)
x2 = tf.placeholder(tf.float32)
x3 = tf.placeholder(tf.float32)

Y = tf.placeholder(tf.float32)

w1 = tf.Variable(tf.random_normal([1]), name='weight1')
w2 = tf.Variable(tf.random_normal([1]), name='weight2')
w3 = tf.Variable(tf.random_normal([1]), name='weight3')
b = tf.Variable(tf.random_normal([1]), name='bias')

hypothesis = x1*w1 + x2*w2 + x3*w3 + b

In [ ]:
# cost function
cost = tf.reduce_mean(tf.square(hypothesis - Y))

# Minimize
optimizer = tf.train.GradientDescentOptimizer(learning_rate=1e-5)
train = optimizer.minimize(cost)

In [ ]:
# Launch the graph
sess = tf.Session()
sess.run(tf.global_variables_initializer())
for step in range(2001):
  cost_val, hy_val, _ = sess.run([cost, hypothesis, train],
                              feed_dict={x1: x1_data, x2: x2_data, x3: x3_data, Y: y_data})
  if step % 10 == 0:
    print(step, "Cost: ", cost_val, "\nPrediction Y: ", hy_val)

0 Cost:  105736.48 
Prediction Y:  [-137.09172  -162.13397  -161.07233  -175.66875  -122.836334]
10 Cost:  1.1914313 
Prediction Y:  [150.55977 183.60208 179.58711 195.30074 140.87338]
20 Cost:  0.22343862 
Prediction Y:  [151.42986 184.6483  180.61777 196.423   141.67157]
30 Cost:  0.22337763 
Prediction Y:  [151.43227 184.65166 180.62083 196.42628 141.67432]
40 Cost:  0.22333081 
Prediction Y:  [151.43202 184.65182 180.62077 196.42613 141.6746 ]
50 Cost:  0.2232792 
Prediction Y:  [151.4318  184.652   180.62071 196.426   141.67493]
60 Cost:  0.22322264 
Prediction Y:  [151.43158 184.65218 180.62067 196.42583 141.67523]
70 Cost:  0.22317958 
Prediction Y:  [151.43135 184.65234 180.62062 196.4257  141.67554]
80 Cost:  0.22312593 
Prediction Y:  [151.43112 184.6525  180.62054 196.42555 141.67584]
90 Cost:  0.22308251 
Prediction Y:  [151.4309  184.65265 180.6205  196.42542 141.67615]
100 Cost:  0.2230274 
Prediction Y:  [151.43066 184.65283 180.62042 196.42528 141.67645]
110 Cost:  0.22

| $x_{1}$ | $x_{2}$ | $x_{3}$ | Y | Pred.Y |
| - | - | - | - | - |
| 73 | 80 | 75 | 152 | 151.39879 |
| 93 | 88 | 93 | 185 | 184.67659 |
| 89 | 91 | 90 | 180 | 180.61304 |
| 96 | 98 | 100 | 196 | 196.40102 |
| 73 | 66 | 70 | 142 | 141.72368 |

> 아주 근사한 근사값을 얻었다! 굿!

그러나 변수가 100개가 넘는 둥 많아지면 셀 [5]을 작성하는 것이 대재앙이 된다. 이제 저렇게 안쓰고 matrix로 아래와 같이 코드를 짜자.



In [ ]:
# Give the data as matrix.
x_data = [[73,	80, 75], [93,	88,	93], [89,	91,	90], [96,	98,	100], [73,	66,	70]]
y_data = [[152],[185],[180],[196],[142]]

# Placeholders for a tensor that will be always fed.
X = tf.placeholder(tf.float32, shape=[None, 3])
Y = tf.placeholder(tf.float32, shape=[None, 1])

W = tf.Variable(tf.random_normal([3, 1]), name='weight')
b = tf.Variable(tf.random_normal([1]), name='bias')

hypothesis = tf.matmul(X, W) + b

# **Loading data from file**

이번 챕터를 진행하기 위해 data-01-test-score.csv를 메모장으로 작성하자.

> **주의!** 엑셀을 사용 시 각 row마다 " "가 붙는다. 이러면 tf.float32로 읽는 데에 지장이 간다.

colab에서의 업로드 & 다운로드 방법은 아래 사이트를 참고하자.

> https://colab.research.google.com/github/StillWork/ds-lab/blob/master/tip_colab_%ED%8C%8C%EC%9D%BC%EC%A0%80%EC%9E%A5%EB%B0%8F%EC%97%85%EB%A1%9C%EB%93%9C_colab.ipynb

2D-array slicing 에 대해서는 아래 사이트를 참고하자.

> https://www.w3schools.com/python/numpy/numpy_array_slicing.asp



In [ ]:
import numpy as np

In [ ]:
# Upload csv file
from google.colab import files
files.upload()

Saving data-01-test-score.csv to data-01-test-score.csv


{'data-01-test-score.csv': b'73,80,75,152\r\n93,88,93,185\r\n89,91,90,180\r\n96,98,100,196\r\n73,66,70,142\r\n53,46,55,101\r\n'}

In [ ]:
# Check a uploaded file successfully
!ls

data-01-test-score.csv	sample_data


In [ ]:
xy = np.loadtxt('data-01-test-score.csv', delimiter=',', dtype=np.float32)
x_data = xy[:, 0:-1]
y_data = xy[:, [-1]]

# Make sure the shape & data are OK
print(x_data, "\nShape: ", x_data.shape, "\nLength: ", len(x_data), "\n")
print(y_data, "\nShape: ", y_data.shape, "\nLength: ", len(y_data), "\n")


[[ 73.  80.  75.]
 [ 93.  88.  93.]
 [ 89.  91.  90.]
 [ 96.  98. 100.]
 [ 73.  66.  70.]
 [ 53.  46.  55.]] 
Shape:  (6, 3) 
Length:  6 

[[152.]
 [185.]
 [180.]
 [196.]
 [142.]
 [101.]] 
Shape:  (6, 1) 
Length:  6 



In [ ]:
# Placeholders for a tensor that will be always fed.
X = tf.placeholder(tf.float32, shape=[None, 3])
Y = tf.placeholder(tf.float32, shape=[None, 1])

W = tf.Variable(tf.random_normal([3, 1]), name='weight')
b = tf.Variable(tf.random_normal([1]), name='bias')

# Hypothesis
hypothesis = tf.matmul(X, W) + b

# Simplified cost function
cost = tf.reduce_mean(tf.square(hypothesis - Y))


In [ ]:
# Minimize
optimizer = tf.train.GradientDescentOptimizer(learning_rate=1e-5)
train = optimizer.minimize(cost)

In [ ]:
# Launch the graph in a session
sess = tf.Session()

In [ ]:
# Initialize global variables in the graph
sess.run(tf.global_variables_initializer())

# Set up feed_dict variables inside the loop
for step in range(2001):
  cost_val, hy_val, _ = sess.run(
      [cost, hypothesis, train],
      feed_dict={X: x_data, Y: y_data})
  if step % 10 == 0:
    print(step, "Cost: ", cost_val, "\nPrediction: ", hy_val)

0 Cost:  22454.814 
Prediction:  [[ 9.730514]
 [16.195074]
 [13.857019]
 [12.041498]
 [16.056774]
 [ 7.794464]]
10 Cost:  4.7062087 
Prediction:  [[149.42683]
 [184.13594]
 [179.31265]
 [192.2262 ]
 [144.15614]
 [102.22739]]
20 Cost:  3.661501 
Prediction:  [[150.37999 ]
 [185.26974 ]
 [180.43596 ]
 [193.44885 ]
 [145.01848 ]
 [102.860695]]
30 Cost:  3.6465862 
Prediction:  [[150.39345 ]
 [185.27385 ]
 [180.44627 ]
 [193.45946 ]
 [145.0191  ]
 [102.858765]]
40 Cost:  3.6318 
Prediction:  [[150.40056]
 [185.27032]
 [180.44908]
 [193.46188]
 [145.01395]
 [102.85255]]
50 Cost:  3.617099 
Prediction:  [[150.40756]
 [185.26675]
 [180.4518 ]
 [193.46423]
 [145.00873]
 [102.84632]]
60 Cost:  3.6025116 
Prediction:  [[150.41457]
 [185.26318]
 [180.45451]
 [193.46657]
 [145.00354]
 [102.84013]]
70 Cost:  3.588023 
Prediction:  [[150.42155]
 [185.25966]
 [180.45723]
 [193.4689 ]
 [144.99838]
 [102.83396]]
80 Cost:  3.5735998 
Prediction:  [[150.42851 ]
 [185.25612 ]
 [180.45993 ]
 [193.47122 ]
 

In [ ]:
# Ask my score
print("Your socre will be ", sess.run(hypothesis, feed_dict={X: [[100, 70, 101]]}))
print("Other socres will be below. \n", sess.run(hypothesis, feed_dict={X: [[60, 70, 110], [90, 100, 80]]}))


Your socre will be  [[182.28418]]
Other socres will be below. 
 [[108.84413]
 [192.23087]]


# **Queue runners**

업로드할 파일이 1개이면 좋겠지만, 다수가 되면 메모리를 초과할 수 있다. 

그래서 tf에서는 파일들을 queue로 적재해 쌓아놓고(batch) 필요할 때마다 꺼내쓰도록 해주는 기능이 있다. \
큰 데이터를 운용할 때 도움된다.

In [ ]:
## Make the lists of files
# filename_queue = tf.train.string_input_producer(
#     ['filename_01.csv', 'filename_02.csv', 'filename_03.csv', ...],
#     shuffle=False, name='filename_queue')

In [ ]:
## Define reader for txt file
# reader = tf.TextLineReader()
# key, value = reader.read(filename_queue)

In [ ]:
## Value parsing
# record_defaults = [[0.], [0.], [0.], [0.]]
# xy = tf.decode_csv(value, record_defaults=record_defaults)

In [ ]:
## Collect batches of csv in
# train_x_batch, train_y_batch = \
# tf.train.batch([xy[0:-1], xy[-1:]], batch_size=10)

In [ ]:
# sess = tf.Session()
# ...

In [ ]:
## Start populating the filename queue
# coord = tf.train.Coordinator()
# threads = tf.train.start_queue_runners(sess=sess, coord=coord)

# for step in range(2001):
#   x_batch, y_batch = sess.run([train_x_batch, train_y_batch])
#   ... feed_dict={X: }

# coord.request_stop()
# coord.join(threads)

Loading data from file의 예시를 Queue runner 방식으로 재현해보자.

In [ ]:
# Make the lists of files
filename_queue = tf.train.string_input_producer(
    ['data-01-test-score.csv'],
    shuffle=False, name='filename_queue')

reader = tf.TextLineReader()
key, value = reader.read(filename_queue)

Instructions for updating:
Queue-based input pipelines have been replaced by `tf.data`. Use `tf.data.TextLineDataset`.


In [ ]:
# Default values, in case of empty columes. Also specifies the type of the decoded result.
record_defaults = [[0.], [0.], [0.], [0.]]      # 데이터 꼴
xy = tf.decode_csv(value, record_defaults=record_defaults)

In [ ]:
# Collect batches of csv in
train_x_batch, train_y_batch = \
tf.train.batch([xy[0:-1], xy[-1:]], batch_size=10)

Instructions for updating:
Queue-based input pipelines have been replaced by `tf.data`. Use `tf.data.Dataset.batch(batch_size)` (or `padded_batch(...)` if `dynamic_pad=True`).


In [ ]:
# Placeholders for a tensor that will be always fed.
X = tf.placeholder(tf.float32, shape=[None, 3])
Y = tf.placeholder(tf.float32, shape=[None, 1])

W = tf.Variable(tf.random_normal([3, 1]), name='weight')
b = tf.Variable(tf.random_normal([1]), name='bias')

hypothesis = tf.matmul(X, W) + b

cost = tf.reduce_mean(tf.square(hypothesis - Y))

In [ ]:
# Minimize
optimizer = tf.train.GradientDescentOptimizer(learning_rate=1e-5)
train = optimizer.minimize(cost)

In [ ]:
# Launch the graph in a session
sess = tf.Session()

In [ ]:
# Initialize global variables in the graph
sess.run(tf.global_variables_initializer())

In [ ]:
# Start populating the filename queue
coord = tf.train.Coordinator()
threads = tf.train.start_queue_runners(sess=sess, coord=coord)

for step in range(2001):
  x_batch, y_batch = sess.run([train_x_batch, train_y_batch])
  cost_val, hy_val, _ = sess.run(
      [cost, hypothesis, train],
      feed_dict={X: x_batch, Y: y_batch}
  )
  if step % 10 == 0:
    print(step, "Cost: ", cost_val, "\nPrediction: ", hy_val)

coord.request_stop()
coord.join(threads)

Instructions for updating:
To construct input pipelines, use the `tf.data` module.
0 Cost:  40678.29 
Prediction:  [[-32.775402]
 [-32.662167]
 [-35.645092]
 [-37.932648]
 [-24.144127]
 [-15.990813]
 [-32.775402]
 [-32.662167]
 [-35.645092]
 [-37.932648]]
10 Cost:  9.6672945 
Prediction:  [[141.18271]
 [105.87815]
 [147.53793]
 [184.08769]
 [177.90794]
 [194.6278 ]
 [141.18271]
 [105.87815]
 [147.53793]
 [184.08769]]
20 Cost:  7.5508814 
Prediction:  [[179.18887 ]
 [196.02008 ]
 [142.16429 ]
 [106.595375]
 [148.62672 ]
 [185.3778  ]
 [179.18887 ]
 [196.02008 ]
 [142.16429 ]
 [106.595375]]
30 Cost:  5.628916 
Prediction:  [[148.53625]
 [185.25049]
 [179.07323]
 [195.89148]
 [142.06476]
 [106.51571]
 [148.53625]
 [185.25049]
 [179.07323]
 [195.89148]]
40 Cost:  8.575914 
Prediction:  [[142.21736]
 [106.62206]
 [148.72052]
 [185.45374]
 [179.2831 ]
 [196.11745]
 [142.21736]
 [106.62206]
 [148.72052]
 [185.45374]]
50 Cost:  7.453284 
Prediction:  [[179.2126 ]
 [196.03798]
 [142.15262]
 [10

In [ ]:
# Ask my score
print("Your socre will be ", sess.run(hypothesis, feed_dict={X: [[100, 70, 101]]}))
print("Other socres will be below. \n", sess.run(hypothesis, feed_dict={X: [[60, 70, 110], [90, 100, 80]]}))


Your socre will be  [[190.46375]]
Other socres will be below. 
 [[163.12186]
 [175.94888]]


Result: 

1.   General
> Your socre will be  [[182.28418]] \
Other socres will be below. 
 [[108.84413]
 [192.23087]]

2.   Queue runner
> Your socre will be  [[190.46375]] \
Other socres will be below. 
 [[163.12186]
 [175.94888]]

조금 엇나간 거같은 느낌이 있지만, 둘 다 나름대로 예측을 해냈다.


여기서 batch들을 순서대로 받아들이는 것에 대해 불만이 있다면 batch를 받을 때 shuffle할 수 있다.

In [ ]:
# min_after_dequeue deines how big a buffer we will randomly sample from
# -- bigger means better shuffling but slower start up and more memory used.

# capacity must be larger than min_after_dequeue and the amount larger determines the maximum we will preftch.
#  Recommandation: 
#       min_after_dequeue + (num_threads + a small safety margin) * batch_size

# min_after_dequeue = 10000
# capacity = min_after_dequeue + 3 * batch_size
# example_batch, label_batch = tf.train.shuffle_batch(
#     [example, label], batch_size=batch_size, capacity=capacity,
#     min_after_dequeue=min_after_dequeue
# )

NameError: ignored